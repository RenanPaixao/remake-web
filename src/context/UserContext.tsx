import { createContext, PropsWithChildren, useEffect, useMemo, useState } from 'react'
import { supabase } from '../utils/supabase.ts'

interface User {
  id: string
  email?: string
  user_metadata: {
    full_name?: string
  }
}

interface IProps{
  user: User | null
  isAuthenticated: boolean
}

export const UserContext = createContext<IProps>({
  user: null,
  isAuthenticated: false
})

UserContext.displayName = 'UserContext'

export const UserProvider = ({ children }: PropsWithChildren) => {
  const [user, setUser] = useState<User | null>(null)

  useEffect(() => {
    supabase.auth.onAuthStateChange((_event, session) => {
      if (session?.user) {
        setUser(session.user)
        return
      }

      setUser(null)
    })

    /**
     * Sets the user by the session.
     */
    async function setUserBySession(): Promise<void> {
      const { data, error } = await supabase.auth.getSession()
      if (error) {
        console.error(error)
        return
      }

      setUser(data.session?.user ?? null)
      return
    }

    setUserBySession()
  }, [])

  /**
   * Returns whether the user is authenticated.
   */
  const isAuthenticated = useMemo(() => {
    const projectId = import.meta.env.VITE_SUPABASE_PROJECT_ID
    const localStorageKey = `sb-${projectId}-auth-token`

    return user !== null || localStorage.getItem(localStorageKey) !== null
  }, [user])

  return <UserContext.Provider value={{
    user,
    isAuthenticated
  }}>
    {children}
  </UserContext.Provider>
}
