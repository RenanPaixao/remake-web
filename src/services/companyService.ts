import { supabase } from '../utils/supabase.ts'
import { Location, LocationsService } from './locationService.ts'

export interface Company {
  id: string
  name: string
  created_at?: string
  owner_id: string
}

export interface CompanyWithLocations extends Company {
  locations: Location[]
}

class PlacesServiceImp {
  get queryBuilder() {
    return supabase.from('companies')
  }
  async getAllWithLocations(options?: {from:number, to:number}): Promise<{ count: number, companies: CompanyWithLocations[] }> {
    const { data: companies, count } = options ?
      await this.queryBuilder.select('*, locations(*)', { count: 'exact' }).range(options.from, options.to).throwOnError() :
      await this.queryBuilder.select('*, locations(*)', { count: 'exact' }).throwOnError()

    if (companies === null) {
      return {
        companies: [],
        count: 0
      }
    }

    return {
      companies: companies.filter(company => company.locations.length),
      count: count ?? 0
    }
  }

  async searchCompanies(query: string): Promise<{ count: number, companies: CompanyWithLocations[] }> {
    const { data: companies, count } = await this.queryBuilder
      .select('*, locations(*)', { count: 'exact' })
      .textSearch('name', `${query}`, {
        type: 'websearch'
      })
      .throwOnError()

    if (companies === null) {
      return {
        companies: [],
        count: 0
      }
    }

    return {
      companies: companies.filter(company => company.locations.length),
      count: count ?? 0
    }
  }

  async deleteCompany(id: string): Promise<void> {
    await this.queryBuilder.delete().eq('id', id).throwOnError()
  }

  async createCompanyWithLocation(
    company: Omit<Company, 'id'>,
    location: Omit<Location, 'company_id'>): Promise<void> {
    const { data, error } = await this.queryBuilder.insert(company).select().maybeSingle()

    if(error) {
      throw new Error(error.message)
    }

    try {
      await LocationsService.create({
        ...location,
        company_id: data.id
      })
    }catch(e: any) {
      await this.deleteCompany(data.id)

      if(e.message) {
        throw new Error(e.message)
      }
      throw new Error('Error creating company')
    }
  }

  async getUserPlaces(userId: string): Promise<CompanyWithLocations[]> {
    const { data: companies } = await this.queryBuilder
      .select('*, locations(*)')
      .eq('owner_id', userId)
      .throwOnError()

    if (companies === null) {
      return []
    }

    return companies.filter(company => company.locations.length)
  }
}

export const CompanyService = new PlacesServiceImp()
