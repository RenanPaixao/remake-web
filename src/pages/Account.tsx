import {
  Avatar,
  Button,
  Card,
  CardBody,
  CardFooter,
  CardHeader,
  Center, Flex,
  Heading,
  Spinner,
  Text
} from '@chakra-ui/react'
import { UserContext } from '../context/UserContext.tsx'
import { useContext, useEffect, useState } from 'react'
import { CompanyService, CompanyWithLocations } from '../services/companyService.ts'
import { PlaceCard } from '../components/PlaceCard/PlaceCard.tsx'
import { showLocationOnMaps } from '../utils/location.ts'
import { useNavigate } from 'react-router-dom'

export const Account = () => {
  const { userInformation } = useContext(UserContext)
  const [isLoading, setIsLoading] = useState(true)
  const [companies, setCompanies] = useState<CompanyWithLocations[]>([])
  const navigate = useNavigate()

  const { id: userId, email, fullName, isRecycler } = userInformation

  useEffect(() => {
    (async() => {
      setIsLoading(true)

      if(userId) {
        const companies = await CompanyService.getUserPlaces(userId)
        setCompanies(companies)
      }
      setIsLoading(false)
    })()
  }, [userId, userInformation])

  /**
   * Render the owned places by the user.
   */
  function renderPlaces() {
    if(companies.length === 0) {
      return <Center>
        <Text>
          You don't have any places yet.
        </Text>
      </Center>
    }

    return companies.map(company => {
      return company.locations.map(location => {
        return <PlaceCard
          key={location.id}
          flexBasis={'15rem'}
          title={company.name}
          distance={''}
          address={`${location.name}, ${location.city}, ${location.state}`}
          onClick={() => showLocationOnMaps({ latitude: location.latitude, longitude: location.longitude })}
        />
      })
    })
  }

  /**
   * Navigate to the add place page.
   */
  function goToAddPlace() {
    navigate('/add-place')
  }

  return <Center py={10} w={'100%'} >
    <Card
      flexBasis={'41rem'}
      boxShadow='rgba(0, 0, 0, 0.56) 0px 22px 70px 4px'
      p={10}
    >
      <CardHeader>
        <Avatar size="2xl" name={fullName}  />
      </CardHeader>
      <CardBody>
        <Heading as="h1" size="xl">
          {fullName}
        </Heading>
        <Text>
          {email}
        </Text>

        <Heading as="h2" size="md" mt={20} mb={10}>
            Your places
        </Heading>

        {isLoading ? <Center>
          <Spinner
            aria-label={'loading'}
            thickness='8px'
            emptyColor={'gray.200'}
            color={'blue.500'}
            speed='0.65s'
            height={50}
            width={50}
          />
        </Center> :
          <Flex
            flexWrap={'wrap'}
            gap={4}
            justifyContent={'center'}
          >
            {renderPlaces()}
          </Flex>
        }
      </CardBody>
      <CardFooter justifyContent={'center'}>
        {isRecycler  ?
          <Button colorScheme={'blue'} onClick={goToAddPlace}>Add New Place</Button> :
          <Button colorScheme={'blue'}>Become a recycler</Button>
        }
      </CardFooter>
    </Card>
  </Center>
}
