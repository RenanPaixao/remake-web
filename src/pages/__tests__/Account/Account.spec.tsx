import { customRender, user } from '../../../../tests/test-utils.tsx'
import { CompanyService } from '../../../services/companyService.ts'
import { waitFor } from '@testing-library/dom'
import { expect } from 'vitest'
import { ReactElement } from 'react'
import { UserContext } from '../../../context/UserContext.tsx'
import { Account } from '../../Account.tsx'
import { ACCOUNT_COMPANIES } from './mock.ts'

const windowOpenMock = vi.fn()
window.open = windowOpenMock

const getUserPlacesMock = vi.fn()
const useNavigateMock = vi.fn()

const mockCompanies = (companies: typeof ACCOUNT_COMPANIES) => {
  CompanyService.getUserPlaces = getUserPlacesMock.mockImplementation(async () => {
    setTimeout(() => {
      return Promise.resolve(companies)
    }, 2000)
  }).mockResolvedValue(companies)
}

const renderWithUserContext = (ui: ReactElement) => {
  return customRender(
    <UserContext.Provider value={{
      userInformation: {
        id: '39f35111-a2d7-48c2-8b0a-ed0a70ed79c6',
        email: 'test.email@email.com',
        fullName: 'john doe',
        isRecycler: true
      },
      isAuthenticated: true
    }}>
      {ui}
    </UserContext.Provider>
  )
}

describe('Account', () => {
  beforeEach(() => {
    vi.clearAllMocks()

    vi.mock('react-router-dom', async importOriginal => {
      const actual: object = await importOriginal()
      return {
        ...actual,
        useNavigate: () => useNavigateMock
      }
    })
  })
  it('should render', async () => {
    mockCompanies(ACCOUNT_COMPANIES)
    const { container, getByRole } = renderWithUserContext(<Account/>)

    const loading = getByRole('generic', { name: 'loading' })

    await waitFor(() => {
      expect(loading).not.toBeInTheDocument()
    })

    expect(container).toMatchSnapshot()
  })

  it('should render when have no companies', async () => {
    mockCompanies([])
    const { container, getByRole } = renderWithUserContext(<Account/>)

    const loading = getByRole('generic', { name: 'loading' })

    await waitFor(() => {
      expect(loading).not.toBeInTheDocument()
    })

    expect(container).toMatchSnapshot()
  })

  it('should call actions from place card and add new place', async () => {
    mockCompanies([ACCOUNT_COMPANIES[0]])

    const { getByRole } = renderWithUserContext(<Account/>)

    const addPlaceButton = getByRole('button', { name: /add new place/i })
    await user.click(addPlaceButton)

    await waitFor(() => {
      expect(useNavigateMock).toHaveBeenCalledWith('/add-place')
      expect(windowOpenMock).not.toHaveBeenCalled()
    })

    const goButton = getByRole('button', { name: 'Go!' })

    await user.click(goButton)

    expect(windowOpenMock.mock.calls[0][0]).contain('https://www.google.com/maps/')
    expect(useNavigateMock).toHaveBeenCalledTimes(1)
  })
})
