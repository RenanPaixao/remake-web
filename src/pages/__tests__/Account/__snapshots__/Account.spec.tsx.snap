// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Account > should render 1`] = `
<div>
  <div
    class="css-1udqmu9"
  >
    <div
      class="chakra-card css-10ospwz"
    >
      <div
        class="chakra-card__header css-1sl53ol"
      >
        <span
          class="chakra-avatar css-j6purw"
        >
          <div
            aria-label="john doe"
            class="chakra-avatar__initials css-1ebyn6"
            role="img"
          >
            jd
          </div>
        </span>
      </div>
      <div
        class="chakra-card__body css-1idwstw"
      >
        <h1
          class="chakra-heading css-1dklj6k"
        >
          john doe
        </h1>
        <p
          class="chakra-text css-0"
        >
          test.email@email.com
        </p>
        <h2
          class="chakra-heading css-nvrxd0"
        >
          Your places
        </h2>
        <div
          class="css-1jrpzwu"
        >
          <div
            class="chakra-card css-1tcl4hn"
          >
            <div
              class="chakra-card__header css-1fj4epu"
            >
              <h3
                class="chakra-heading css-1q6o3tu"
              >
                testedoteste
              </h3>
              <h5
                class="chakra-text css-1vuos2j"
              />
            </div>
            <div
              class="chakra-card__footer css-1r7igq6"
            >
              <p
                class="chakra-text css-cg4qa7"
              >
                rua, Carpina, pe
              </p>
              <button
                class="chakra-button css-1noe2cj"
                type="button"
              >
                Go!
              </button>
            </div>
          </div>
          <div
            class="chakra-card css-1tcl4hn"
          >
            <div
              class="chakra-card__header css-1fj4epu"
            >
              <h3
                class="chakra-heading css-1q6o3tu"
              >
                Empresa - Derby
              </h3>
              <h5
                class="chakra-text css-1vuos2j"
              />
            </div>
            <div
              class="chakra-card__footer css-1r7igq6"
            >
              <p
                class="chakra-text css-cg4qa7"
              >
                principal teste, recife, pernambuco
              </p>
              <button
                class="chakra-button css-1noe2cj"
                type="button"
              >
                Go!
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="chakra-card__footer css-1l9v0bz"
      >
        <button
          class="chakra-button css-jut409"
          type="button"
        >
          Add New Place
        </button>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;

exports[`Account > should render when have no companies 1`] = `
<div>
  <div
    class="css-1udqmu9"
  >
    <div
      class="chakra-card css-10ospwz"
    >
      <div
        class="chakra-card__header css-1sl53ol"
      >
        <span
          class="chakra-avatar css-j6purw"
        >
          <div
            aria-label="john doe"
            class="chakra-avatar__initials css-1ebyn6"
            role="img"
          >
            jd
          </div>
        </span>
      </div>
      <div
        class="chakra-card__body css-1idwstw"
      >
        <h1
          class="chakra-heading css-1dklj6k"
        >
          john doe
        </h1>
        <p
          class="chakra-text css-0"
        >
          test.email@email.com
        </p>
        <h2
          class="chakra-heading css-nvrxd0"
        >
          Your places
        </h2>
        <div
          class="css-1jrpzwu"
        >
          <div
            class="css-gmuwbf"
          >
            <p
              class="chakra-text css-0"
            >
              You don't have any places yet.
            </p>
          </div>
        </div>
      </div>
      <div
        class="chakra-card__footer css-1l9v0bz"
      >
        <button
          class="chakra-button css-jut409"
          type="button"
        >
          Add New Place
        </button>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
